### PR TITLE
crypto: fix BackedUpRoomKey Serialization

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/3247

A source of https://github.com/matrix-org/matrix-rust-sdk/issues/3197

When downloading keys if the `forwarding_curve25519_key_chain` contains items the deserialisation would fail with `error: Json(Error("invalid type: string \"EQCtWSJclhxEU9mGlF4erDHxIacpUQz70yjYVRPODTg\", expected an array of length 32", line: 1, column: 255))`

As a consequence the key import fails silently with:
`Successfully imported room keys total_count=0 imported_count=0 room_keys={} `
That would cause unable to decrypt errors for history.

A serde attribute `deserialize_curve_key_vec`  was missing.

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/olm/group_sessions/mod.rs [hunk 1]
```diff
@@ -245,3 +250,28 @@ impl TryFrom<ForwardedRoomKeyContent> for ExportedRoomKey {
...
+                "session_id": "/2K+V777vipCxPZ0gpY9qcpz1DYaXwuMRIu0UEP0Wa0",
+                "session_key": "AQAAAAAclzWVMeWBKH+B/WMowa3rb4ma3jEl6n5W4GCs9ue65CruzD3ihX+85pZ9hsV9Bf6fvhjp76WNRajoJYX0UIt7aosjmu0i+H+07hEQ0zqTKpVoSH0ykJ6stAMhdr6Q4uW5crBmdTTBIsqmoWsNJZKKoE2+ldYrZ1lrFeaJbjBIY/9ivle++74qQsT2dIKWPanKc9Q2Gl8LjESLtFBD9Fmt",
+                "sender_claimed_keys": {
+                    "ed25519": "F4P7f1Z0RjbiZMgHk1xBCG3KC4/Ng9PmxLJ4hQ13sHA"
+                },
+                "forwarding_curve25519_key_chain": ["DBPC2zr6c9qimo9YRFK3RVr0Two/I6ODb9mbsToZN3Q", "bBc/qzZFOOKshMMT+i4gjS/gWPDoKfGmETs9yfw9430"]
+        });
+
+        let backed_up_room_key: BackedUpRoomKey = serde_json::from_value(data).unwrap();
+        assert_eq!(backed_up_room_key.forwarding_curve25519_key_chain.len(), 2);
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/mod.rs:275]:
```suggestion
        assert_eq!(backed_up_room_key.forwarding_curve25519_key_chain.len(), 2, "The number of forwarding Curve25519 chains should be two.");
```

**Comment by richvdh** [comment]
AFAICT this is a fix to *de*serialization, not serialization. I have updated the PR title accordingly.

**Comment by poljar** [comment]
> AFAICT this is a fix to _de_serialization, not serialization. I have updated the PR title accordingly.

It's for both:

image

**Comment by richvdh** [comment]
Hrm ok. Serialisation doesn't appear to be tested; and the bug it clams to fix (#3247), and the description of the PR, only mention *de*serialisation. Hence my confusion.

Any idea what the effect of this issue would have been on serialization? Presumably the `forwarding_curve25519_key_chain` property would have been serialised as an array of individual bytes, or something? But maybe that didn't really matter in practice because we so rarely use room keys with `forwarding_curve25519_key_chain`?

**Comment by poljar** [comment]
> Any idea what the effect of this issue would have been on serialization? Presumably the `forwarding_curve25519_key_chain` property would have been serialised as an array of individual bytes, or something? But maybe that didn't really matter in practice because we so rarely use room keys with `forwarding_curve25519_key_chain`?

Yes, all of that is correct. It would not base64 encode the keys, it would instead create a 32 byte long array. I'm not completely sure, but I think it didn't matter as the SDK doesn't really use the forwarding chain, I would need to look at the codepaths where we receive room keys.

**Comment by richvdh** [comment]
> As a consequence the key import fails silently with:

For links: the fact that deserialization errors were swallowed was a separate problem, seemingly fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/3149.

**Pull request merged**